worker/credentialvalidator/util_test.go

```diff
@@ -0,0 +1,155 @@
+		return nil, err
+	}
+	return newMockWatcher(), nil
+}
+
+// newMockWatcher returns a watcher.NotifyWatcher that always
+// sends 3 changes and then sits quietly until killed.
+func newMockWatcher() *mockWatcher {
+	const count = 3
+	changes := make(chan struct{}, count)
+	for i := 0; i < count; i++ {
+		changes <- struct{}{}
+	}
+	return &mockWatcher{
+		Worker:  workertest.NewErrorWorker(nil),
+		changes: changes,
+	}
+}
+
+// mockWatcher implements watcher.NotifyWatcher for use in the tests.
+type mockWatcher struct {
+	worker.Worker
+	changes chan struct{}
+}
+
+// Changes is part of the watcher.NotifyWatcher interface.
+func (mock *mockWatcher) Changes() watcher.NotifyChannel {
+	return mock.changes
+}
+
+// modelUUID is the model tag we're using in the tests.
+var modelUUID = "01234567-89ab-cdef-0123-456789abcdef"
+
+// credentialTag is the credential tag we're using in the tests.
+// needs to fit fmt.Sprintf("%s/%s/%s", cloudName, userName, credentialName)
+var credentialTag = names.NewCloudCredentialTag("cloud/user/credential").String()
+
+// panicCheck is a Config.Check value that should not be called.
+func panicCheck(base.StoredCredential) bool { panic("unexpected") }
+
+// neverCheck is a Config.Check value that always returns false.
+func neverCheck(base.StoredCredential) bool { return false }
+
+// panicFacade is a NewFacade that should not be called.
+func panicFacade(base.APICaller) (credentialvalidator.Facade, error) {
+	panic("panicFacade")
+}
+
+// panicWorker is a NewWorker that should not be called.
+func panicWorker(credentialvalidator.Config) (worker.Worker, error) {
+	panic("panicWorker")
+}
+
+// validConfig returns a minimal config stuffed with dummy objects that
+// will explode when used.
+func validConfig() credentialvalidator.Config {
+	return credentialvalidator.Config{
+		Facade: struct{ credentialvalidator.Facade }{},
+		Check:  panicCheck,
+	}
+}
+
+// checkNotValid checks that the supplied credentialvalidator.Config fails to
+// Validate, and cannot be used to construct a credentialvalidator.Worker.
+func checkNotValid(c *gc.C, config credentialvalidator.Config, expect string) {
+	check := func(err error) {
+		c.Check(err, gc.ErrorMatches, expect)
+		c.Check(err, jc.Satisfies, errors.IsNotValid)
+	}
+
+	err := config.Validate()
+	check(err)
+
+	worker, err := credentialvalidator.New(config)
+	c.Check(worker, gc.IsNil)
+	check(err)
+}
+
+// validManifoldConfig returns a minimal config stuffed with dummy objects
+// that will explode when used.
+func validManifoldConfig() credentialvalidator.ManifoldConfig {
+	return credentialvalidator.ManifoldConfig{
+		APICallerName: "api-caller",
+		Check:         panicCheck,
+		NewFacade:     panicFacade,
+		NewWorker:     panicWorker,
+	}
+}
+
+// checkManifoldNotValid checks that the supplied ManifoldConfig creates
+// a manifold that cannot be started.
+func checkManifoldNotValid(c *gc.C, config credentialvalidator.ManifoldConfig, expect string) {
```

Comment:
Interesting approach. In other workers, we use an exported Validate() method on the config and just call that directly. There's a bit more boilerplate with this approach.

Reply:
i followed whatever was in migration flag. I am not big fun of exporting functions that are not otherwise exported just for testing...

Reply:
what function would you be exporting that isn't otherwise exported?

Reply:
Sure, ok. Can go either way with this one. We sometimes export methods to make testing easier. But this way is fine, was just thinking out loud. I wish we didn't have so many ways things that are the same have been implemented over time.

Reply:
I don't think it's worth putting the foot down. I'll export Validate() and will change the test to call it directly rather than going through Start().

I do not see a significant difference in implementation - nothing gained through exporting it from behavioral or performance consideration. 
However, I agree that the test is simpler and leaner and this can be a metrics of quality by itself :D